src/sentry/options/store.py

```diff
@@ -90,19 +174,61 @@ def set_store(self, key, value):
         )
 
     def set_cache(self, key, value):
-        self.cache.set(key.cache_key, value, self.ttl)
-
-    def delete(self, key):
-        self.delete_store(key)
+        cache_key = key.cache_key
+        if key.ttl > 0:
+            self._local_cache[cache_key] = _make_cache_value(key, value)
         try:
-            self.delete_cache(key)
+            self.cache.set(cache_key, value, self.ttl)
             return True
         except Exception:
-            logger.warn(CACHE_UPDATE_ERR, key, exc_info=True)
+            logger.warn(CACHE_UPDATE_ERR, key.name, exc_info=True)
             return False
 
+    def delete(self, key):
+        """
+        Remove key out of option stores. This operation must succeed on the
+        database first. If database fails, an exception is raised.
+        If database succeeds, caches are then allowed to fail silently.
+        A boolean is returned to indicate if the network deletion succeeds.
+        """
+        self.delete_store(key)
+        return self.delete_cache(key)
+
     def delete_store(self, key):
         Option.objects.filter(key=key.name).delete()
 
     def delete_cache(self, key):
-        self.cache.delete(key.cache_key)
+        cache_key = key.cache_key
+        try:
+            del self._local_cache[cache_key]
+        except KeyError:
+            pass
+        try:
+            self.cache.delete(cache_key)
+            return True
+        except Exception:
+            logger.warn(CACHE_UPDATE_ERR, key, exc_info=True)
+            return False
+
+    def expire_local_cache(self):
+        """
+        Iterate over our local cache items, and
+        remove the keys that are beyond their grace time.
+        """
+        if not self._local_cache:
+            return
+
+        to_expire = []
+        now = int(time())
+        for k, (_, _, grace) in self._local_cache.iteritems():
+            if now > grace:
+                to_expire.append(k)
+
+        for k in to_expire:
+            del self._local_cache[k]
+
+    def flush_local_cache(self):
+        """
+        Empty store's local in-process cache.
+        """
+        self._local_cache = {}
```

Comment:
this doesnt appear to be thread bound at all, which means any request could flush the local cache

that can probably be ok, but im not sure if thats what you're intending (and also it might need locks, though might not)